src/mlpack/tests/decision_tree_test.cpp

```diff
@@ -1082,4 +1085,45 @@ BOOST_AUTO_TEST_CASE(ConstDataTest)
       constWeights);
 }
 
+/**
+ * Construct the decision tree with splitting only if gain is more than
+ * threshold.
+ */
+BOOST_AUTO_TEST_CASE(RegularisedDecisionTree)
+{
+  // Completely random dataset with no structure.
+  arma::mat dataset(10, 1000, arma::fill::randu);
+  arma::Row<size_t> labels(1000);
+  for (size_t i = 0; i < 1000; ++i)
+    labels[i] = i % 3; // 3 classes.
+  arma::rowvec weights(labels.n_elem);
+  weights.ones();
+
+  // Minimum leaf size of 1.
+  DecisionTree<> d(dataset, labels, 3, weights, 1, 1e-7);
+
+  // Minimum leaf size of 1 and Minimum gain split of 0.01.
+  DecisionTree<> dRegularised(dataset, labels, 3, weights, 1, 0.01);
+
+  size_t count = 0;
+  // This part of code is dupliacte with no weighted one.
+  for (size_t i = 0; i < 1000; ++i)
+  {
+    size_t prediction, predictionsregularised;
+    arma::vec probabilities, probabilitiesRegularised;
+
+    d.Classify(dataset.col(i), prediction, probabilities);
+    dRegularised.Classify(dataset.col(i), predictionsregularised,
+                          probabilitiesRegularised);
+
+    if (prediction != predictionsregularised)
+      count++;
+
+    BOOST_REQUIRE_EQUAL(probabilities.n_elem, 3);
+    BOOST_REQUIRE_EQUAL(probabilitiesRegularised.n_elem, 3);
+  }
+
+  BOOST_REQUIRE_GT(count, 0);
```

Comment:
In this test we need to make sure regularization is working (note that this is different than the main tests, where we only need to make sure that the option is being properly passed to the code), but it looks like this test just checks if the predictions produced by the regularized tree are different than the nonregularized tree.

I think a more effective test might be to make sure that a tree built with a high minimum gain split has fewer nodes in it than a tree with low minimum gain split.  What do you think?

We could also test that the accuracy on a training set of a regularized tree is higher, but I am not sure that will always be true so I might be hesitant about that testing strategy.

In fact for this test here, I think it would be worthwhile to adapt and move it to `main_tests/decision_tree_test.cpp`.

Reply:
I agree that it is not a strict test for regularisation. I considered this hypothesis  :- Since the unregularised tree gives 100% accuracy on training data, the fact that the `count > 0` means that the tree is regualrised and it contains less nodes. Had it contained more nodes, it can't have `count > 0`. So what you said is somehow indirectly tested. Thisis the reason I adopted this strategy.

Reply:
I guess, then this should be good to go?
Any views @rcurtin .

Reply:
Agreed, but I do still think it would be useful to add a similar test to `main_tests/decision_tree_test.cpp` as paranoia just to ensure the regularization is being passed down properly.

Reply:
Sure, I will add it then.